internal/jobservice/eventstojobs/eventstojobs.go

```diff
@@ -4,101 +4,122 @@ import (
+	ctx, cancel := context.WithCancel(ctx)
 	g, _ := errgroup.WithContext(ctx)
+	expiresAt := time.Now().Add(time.Duration(timeout) * time.Second)
+
 	g.Go(func() error {
+		defer cancel()
+
 		// Once we unsubscribed from the job-set, we need to close the GRPC connection.
 		// According to GRPC official docs, you can only end a client stream by either canceling the context or closing the connection
-		// This will log an error to the jobservice log saying that the connection was used.
-		ticker := time.NewTicker(time.Duration(eventToJobService.jobServiceConfig.SubscribeJobSetTime) * time.Second)
-		for range ticker.C {
-			if !eventToJobService.jobServiceRepository.IsJobSetSubscribed(eventToJobService.queue, eventToJobService.jobSetId) {
-				return conn.Close()
+		// // This will log an error to the jobservice log saying that the connection was used.
+		ticker := time.NewTicker(time.Duration(timeout) * time.Second)
+		for {
+			select {
+			case <-ctx.Done():
+				return nil
+			case t := <-ticker.C:
+				if !eventToJobService.jobServiceRepository.IsJobSetSubscribed(
+					eventToJobService.queue,
+					eventToJobService.jobSetId) {
+					return nil
+				}
+				if t.After(expiresAt) {
+					return errors.Errorf("stream subscription ttl exceeded: %v", timeout)
+				}
 			}
 		}
-		return nil
 	})
 	g.Go(func() error {
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		default:
-		}
-		eventClient := api.NewEventClient(conn)
-		stream, err := eventClient.GetJobSetEvents(ctx, &api.JobSetRequest{
-			Id:             eventToJobService.jobSetId,
-			Queue:          eventToJobService.queue,
-			Watch:          true,
-			FromMessageId:  fromMessageId,
-			ErrorIfMissing: false,
-		})
-		if err != nil {
-			log.Errorf("Error found from client %v", err)
-			return err
-		}
-		for {
-
-			msg, err := stream.Recv()
+		var err error
+		defer func() {
+			eventToJobService.eventClient.Close()
+			// cancel the ticker go routine if an error originated here
 			if err != nil {
-				log.Error(err)
-				return err
+				cancel()
 			}
-			fromMessageId = msg.GetId()
-			currentJobId := api.JobIdFromApiEvent(msg.Message)
-			jobStatus := EventsToJobResponse(*msg.Message)
-			if jobStatus != nil {
-				jobTable := repository.NewJobTable(eventToJobService.queue, eventToJobService.jobSetId, currentJobId, *jobStatus)
-				eventToJobService.jobServiceRepository.UpdateJobServiceDb(jobTable)
+		}()
+
+		// this loop will run until the context is canceled or an error is encountered
+		for {
+			select {
+			case <-ctx.Done():
+				return nil
+			default:
+				msg, err := eventToJobService.eventClient.GetJobEventMessage(ctx, &api.JobSetRequest{
+					Id:             eventToJobService.jobSetId,
+					Queue:          eventToJobService.queue,
+					Watch:          true,
+					FromMessageId:  fromMessageId,
+					ErrorIfMissing: false,
+				})
+				if err != nil {
+					log.Error(err)
+					return err
+				}
+				fromMessageId = msg.GetId()
```

Comment:
If you go with my suggestion of handelling the error above, then move the update of `fromMessageId` to after the database insert.  That when when that function returns an error (right now it panics, but it shouldn't) we can only advance our position in the stream after we know the db insert was good.

Reply:
done